### PR TITLE
feature: convert plan item code to Kotlin

### PR DESCRIPTION
**Pull request by hsiliev**
Convert planitem package and related suspension util to Kotlin

Solves: PZ-4367

### CODE REVIEW

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+) {
+
+    companion object {
+        private const val REDEN_OPSCHORTING = "Aanvullende informatie opgevraagd"
+        private const val REDEN_PAST_FATALE_DATUM = "Aanvullende informatie opgevraagd"
+    }
+
+    @GET
+    @Path("zaak/{uuid}/humanTaskPlanItems")
+    fun listHumanTaskPlanItems(@PathParam("uuid") zaakUUID: UUID?): List<RESTPlanItem> =
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:86]:
```suggestion
    fun listHumanTaskPlanItems(@PathParam("uuid") zaakUUID: UUID): List<RESTPlanItem> =
```

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    fun listProcessTaskPlanItems(@PathParam("uuid") zaakUUID: UUID?): List<RESTPlanItem> =
+        cmmnService.listProcessTaskPlanItems(zaakUUID).let { processTaskPlanItems ->
+            zrcClientService.readZaak(zaakUUID).let { zaak ->
+                planItemConverter.convertPlanItems(processTaskPlanItems, zaak)
+            }
+        }
+
+    @GET
+    @Path("zaak/{uuid}/userEventListenerPlanItems")
+    fun listUserEventListenerPlanItems(@PathParam("uuid") zaakUUID: UUID?): List<RESTPlanItem> =
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:106]:
```suggestion
    fun listUserEventListenerPlanItems(@PathParam("uuid") zaakUUID: UUID): List<RESTPlanItem> =
```

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                        planItemConverter.convertPlanItem(planItemInstance, zaakUUID, zaps)
+                    }
+                }
+            }
+        }
+
+    @POST
+    @Path("doHumanTaskPlanItem")
+    @Suppress("LongMethod")
+    fun doHumanTaskplanItem(humanTaskData: @Valid RESTHumanTaskData) {
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:138]: This does not work? The `@Valid` needs to be in front of the parameter, not the type?

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    private fun calculateFatalDate(
+        humanTaskData: RESTHumanTaskData,
+        zaakafhandelParameters: ZaakafhandelParameters,
+        planItem: PlanItemInstance,
+        zaak: Zaak
+    ): LocalDate? {
+        val humanTaskParameters = zaakafhandelParameters.findHumanTaskParameter(planItem.planItemDefinitionId)
+        val zaakFatalDate = zaak.uiterlijkeEinddatumAfdoening
+
+        if (humanTaskData.fataledatum != null) {
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:274]: replace with `?.let{..}`?

File: src/main/kotlin/net/atos/zac/app/planitems/model/DefaultHumanTaskFormulierKoppeling.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+) {
+    AANVULLENDE_INFORMATIE("AANVULLENDE_INFORMATIE", FormulierDefinitie.AANVULLENDE_INFORMATIE),
+    GOEDKEUREN("GOEDKEUREN", FormulierDefinitie.GOEDKEUREN),
+    ADVIES_INTERN("ADVIES_INTERN", FormulierDefinitie.ADVIES),
+    ADVIES_EXTERN("ADVIES_EXTERN", FormulierDefinitie.EXTERN_ADVIES_VASTLEGGEN),
+    DOCUMENT_VERZENDEN_POST("DOCUMENT_VERZENDEN_POST", FormulierDefinitie.DOCUMENT_VERZENDEN_POST),
+    DEFAULT("", FormulierDefinitie.DEFAULT_TAAKFORMULIER);
+
+    companion object {
+        @JvmStatic
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/model/DefaultHumanTaskFormulierKoppeling.kt:22]: Remove `@JvmStatic`? Maybe move these static functions out of the enum altogether? Seems a bit weird now?

File: src/main/kotlin/net/atos/zac/app/planitems/model/RESTHumanTaskData.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import java.time.LocalDate
+
+@NoArgConstructor
+data class RESTHumanTaskData(
+    var planItemInstanceId: String? = null,
+
+    @field:NotNull
+    var groep: RestGroup,
+
+    var medewerker: @Valid RestUser? = null,
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/model/RESTHumanTaskData.kt:21]:
```suggestion
   @Valid
    var medewerker: RestUser? = null,
```

**Review by edgarvonk — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    @GET
+    @Path("zaak/{uuid}/humanTaskPlanItems")
+    fun listHumanTaskPlanItems(@PathParam("uuid") zaakUUID: UUID?): List<RESTPlanItem> =
+        cmmnService.listHumanTaskPlanItems(zaakUUID).let { humanTaskPlanItems ->
+            zrcClientService.readZaak(zaakUUID).let { zaak ->
+                planItemConverter.convertPlanItems(humanTaskPlanItems, zaak)
+                    .filter { it.actief }
+                    .toList()
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:91]: Nope

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt (hunk 3 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:138]: Yeah. You're right. Thanks!

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        val zaakUUID = zaakVariabelenService.readZaakUUID(planItem)
+        val zaak = zrcClientService.readZaak(zaakUUID)
+        val taakdata = humanTaskData.taakdata
+        PolicyService.assertPolicy(policyService.readZaakRechten(zaak).startenTaak)
+        val zaakafhandelParameters = zaakafhandelParameterService.readZaakafhandelParameters(
+            UriUtil.uuidFromURI(zaak.zaaktype)
+        )
+
+        val fatalDate = calculateFatalDate(humanTaskData, zaakafhandelParameters, planItem, zaak)
+        if (fatalDate != null) {
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:149]: Used `also` as let implies that the result from the block would be used and this breaks the IFs below

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                    ),
+                    zaak.getBronnenFromZaak()
+                )
+            )
+        }
+
+        cmmnService.startHumanTaskPlanItem(
+            humanTaskData.planItemInstanceId,
+            humanTaskData.groep.id,
+            if (humanTaskData.medewerker != null && humanTaskData.medewerker.toString().isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:194]: We really do want the string representation fo `medwerker` here. Not sure if we can get rid of the `toString`

File: src/main/kotlin/net/atos/zac/app/planitems/converter/RESTPlanItemConverter.kt [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    tabellen[rt.veld] = rt.tabel.values.map(ReferenceTableValue::name).toList()
+                }
+                groepId = it.groepID
+                if (it.doorlooptijd != null) {
+                    fataleDatum = LocalDate.now().plusDays(it.doorlooptijd.toLong())
+                }
+            }
+        }
+
+    private fun convertProcessTask(restPlanItem: RESTPlanItem): RESTPlanItem =
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/converter/RESTPlanItemConverter.kt:82]: Does not make sense indeed. Removed it + the line 47 that calls it

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        cmmnService.startUserEventListenerPlanItem(userEventListenerData.planItemInstanceId)
+        if (userEventListenerData.restMailGegevens != null) {
+            mailService.sendMail(
+                restMailGegevensConverter.convert(userEventListenerData.restMailGegevens),
+                zaak.getBronnenFromZaak()
+            )
+        }
+    }
+
+    @Suppress("ReturnCount")
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt:263]: Yeah, not very nice those 3 return statements I feel. But oh well, for another day. :-)

**Review by edgarvonk — APPROVED** [review]
nice!

File: src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt (hunk 11 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/planitems/PlanItemsRESTService.kt]: Extracted the code in a new function and now we have only 2 returns, which makes detekt happy

**Pull request merged**